internal/dcache/replication_manager/replication_manager.go

```diff
@@ -399,27 +530,46 @@ retry:
 			}
 
 			log.Debug("ReplicationManager::WriteMV: Sending PutChunk request for %s/%s to node %s: %s",
-				rv.Name, req.MvName, targetNodeID, rpc.PutChunkRequestToString(rpcReq))
+				rvName, req.MvName, targetNodeID, rpc.PutChunkRequestToString(putChunkReq))
 
-			//
-			// Schedule PutChunk RPC call to the target node.
-			// One of the threadpool threads will pick this request and call PutChunk.
-			// Since we have to wait for all the replica writes to complete before we
-			// can start processing the individual responses we send the last replica
-			// inline and save one threadpool thread.
-			//
-			isLastComponentRV := componentRVIdx == (len(componentRVs) - 1)
+			isLastComponentRV := componentRVIdx == (len(putChunkExReq.NextRVs) - 1)
 			rm.tp.schedule(&workitem{
 				targetNodeID: targetNodeID,
-				rvName:       rv.Name,
-				putChunkReq:  rpcReq,
+				rvName:       rvName,
+				putChunkReq:  putChunkReq,
 				respChannel:  responseChannel,
 			}, isLastComponentRV /* runInline */)
+		}
+	} else if PutChunkStyle == DaisyChain {
+		rvName := getRvNameFromRvID(putChunkExReq.Request.Address.RvID)
+		targetNodeID := getNodeIDFromRVName(rvName)
+		common.Assert(common.IsValidUUID(targetNodeID))
+
+		log.Debug("ReplicationManager::WriteMVEx: Sending PutChunkEx request for %s/%s to node %s: %s",
+			rvName, req.MvName, targetNodeID, rpc.PutChunkExRequestToString(putChunkExReq))
+
+		ctx, cancel := context.WithTimeout(context.Background(), RPCClientTimeout*time.Second)
+		defer cancel()
+
+		putChunkExResp, err := rpc_client.PutChunkEx(ctx, targetNodeID, putChunkExReq)
+
+		//
+		// [Sourav]
+		// Write a function that converts putChunkExResp to proper responses on the responseChannel, nil
+		// response for all if PutChunkEx() call failed to send the request.
+		//
+		if err != nil {
+			log.Err("ReplicationManager::WriteMVEx: Failed to send PutChunkEx request for %s/%s to node %s: %v",
+				rv.Name, req.MvName, targetNodeID, err)
+			common.Assert(putChunkExResp == nil)
 		} else {
-			common.Assert(false, "Unexpected RV state", rv.State, rv.Name, req.MvName)
+			log.Debug("ReplicationManager::WriteMVEx: Received PutChunkEx response for %s/%s from node %s: %s",
+				rv.Name, req.MvName, targetNodeID, rpc.PutChunkExResponseToString(putChunkExResp))
+			common.Assert(putChunkExResp != nil)
 		}
 	}
 
+processResponses:
```

Comment:
you will need to handle ThriftError also, treating it as transport error from the nexthop RV